url2b9f720f64_jfimbett_QBnB_tgz-pJ8-tests_TestJ8/src/algorithms/SimplexAlgorithm.java

```diff
@@ -0,0 +1,1063 @@
+	  double[][] fin=new double[X.length][X[0].length-1];
+	  for (int i = 0; i < fin.length; i++) {
+	   int j=0;
+	   int k=0;
+	   while(k<X[0].length){
+	    if(k!=pos){
+	     fin[i][j]=X[i][k];
+	     j++;
+	    }
+	    k++;
+	   }
+	  }
+	  return fin;
+	 }
+	 
+	 /**
+	  * Adds a new column to a matrix
+	  * @param n: new column
+	  * @param X: The Matrix
+	  * @return X+{n}
+	  */
+	 public double[][] addColumnToMatrix(double[] n, double[][] X){
+	  double[][] fin= new double[X.length][X[0].length+1];
+	  for (int i = 0; i < fin.length; i++) {
+	   for (int j = 0; j < fin[0].length; j++) {
+	    if(j==fin[0].length-1){
+	     fin[i][j]=n[i];
+	    }
+	    else{
+	     fin[i][j]=X[i][j];
+	    }
+	    
+	   }
+	  }
+	  return fin;
+	 }
+	 
+	 /**
+	  * Adds element to the end of a vector
+	  * @param val: The element to be added
+	  * @param vector: The vector
+	  * @return The original vector plus val at the end
+	  */
+	 public double[] addToVector(double val, double[] vector){
+	  double[] newVector=new double[vector.length+1];
+	  for (int i = 0; i < newVector.length; i++) {
+	   if(i<vector.length){
+	    newVector[i]=vector[i];
+	   }
+	   else{
+	    newVector[i]=val;
+	   }
+	  }
+	  return newVector;
+	 }
+	 
+	 
+
+	 /**
+	  * Adds element to the end of a vector
+	  * @param val: The element to be added
+	  * @param vector: The vector
+	  * @return The original vector plus val at the end
+	  */
+	 public String[] addToVector(String val, String[] vector){
+	  String[] newVector=new String[vector.length+1];
+	  for (int i = 0; i < newVector.length; i++) {
+	   if(i<vector.length){
+	    newVector[i]=vector[i];
+	   }
+	   else{
+	    newVector[i]=val;
+	   }
+	  }
+	  return newVector;
+	 }
+	 
+	 /**
+	  * Returns the primal objective function
+	  * @return
+	  */
+	 public double primalfo(){
+	  return dotProduct(cb, xb)+dotProduct(cn, xn);
+	 }
+	 
+	 /**
+	  * Creates a feasible solution if the Identity Matrix is not found on A
+	  * @param A: The matrix of coefficients
+	  * @param c: The cost vector
+	  * @param I: The variable names
+	  */
+	 public void createBasis(double[][] A, double[] c, String[] I){
+	  //Identifies how many artificial variables are needed
+	  //Finds an identity matrix in A
+	  //Indicates in which columns are the necessary columns to form a base
+	  B= new double[0][0];
+	  int numArt=0;
+	  for (int i = 0; i < A.length; i++) {
+	   double[] Ii= new double[A.length];
+	   Ii[i]=1;
+	   int pos=isColumn(Ii, A);
+	   if(pos==-1){
+	    //Adds artificial variable
+	    String idAr="a"+numArt;
+	    nart++;
+	    numArt++;
+	    Ib=addToVector(idAr, Ib);
+	    //Depends on the initialization methodology
+	    if(true){
+	    cb=addToVector(99999999, cb);
+	    }
+	    double[] column=Ii;
+	    if(B.length!=0){
+	     B=addColumnToMatrix(column, B);
+	    }
+	    else{
+	     B=new double[Ii.length][1];
+	     for (int j = 0; j < Ii.length; j++) {
+	      B[j][0]=Ii[j];
+	     }
+	    } 
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 368:376, 395:403.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.